File: Modulo6ProcesamientoDatosEscalable/Tareas/ProyectoPythonSpark/src/SpanishAirports.py
"""

Original file is located at
    https://colab.research.google.com/drive/1wf0YLhk4cGSWDyBUuGYElxkUXrQcOBna

"
Reads the airports.csv file and counts the number of Spanish airports
for each type, writing the results in a text file.
""
"""

from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession

# Configurar Spark

spark = SparkSession.builder \
    .appName("MyApp") \
    .config("spark.driver.memory", "4g") \
    .config("spark.executor.memory", "4g") \
    .getOrCreate()

spark_context =  spark.sparkContext
spark_context.setLogLevel("ERROR")

# Leer el archivo
path_file = "C:/Users/cberd/Downloads/airports.csv"

rdd = spark_context.textFile(path_file).cache()


# Obtener la cabecera y limpiarla de comillas
header = rdd.first()


column_names = header.split(",")
print(column_names)

# Obtener los índices correctos

index_iso_country = column_names.index('"iso_country"')
index_type = column_names.index('"type"')

filtered_rdd = (
    rdd
    .filter(lambda row: row != header)
    .map(lambda row: row.split(","))
    .filter(lambda cols: cols[index_iso_country] == '"ES"')
)

# Mapear cada fila al tipo de aeropuerto como clave y contar
type_counts = (
    filtered_rdd
    .map(lambda cols: (cols[index_type], 1))  # Crear pares clave-valor (type, 1)
    .reduceByKey(lambda a, b: a + b)  # Sumar los valores por clave
    .collect()  # Traer el resultado a Python
)

# Convertir el resultado en un diccionario para mejor visualización
type_counts_dict = dict(type_counts)

# Imprimir los conteos
print(type_counts_dict)

# Guardar resultados en un archivo
with open("result.txt", "w") as file:
    file.write(str(type_counts_dict))


spark_context.stop()

File: Modulo6ProcesamientoDatosEscalable/Tareas/ProyectosJavaSpark/AddNumbersSpark/src/main/java/spark/SpanishAirports.java
package spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import java.io.FileWriter;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class SpanishAirports {

    public static void writeResult(long smallAirport, long heliport, long mediumAirport, long closed, long largeAirport, long seaplaneBase) {
        // Guardar resultados en un archivo
        try (FileWriter writer = new FileWriter("./src/main/resources/result.txt")) {
            writer.write("small_airport: " + smallAirport + "\n");
            writer.write("heliport: " + heliport + "\n");
            writer.write("medium_airport: " + mediumAirport + "\n");
            writer.write("closed: " + closed + "\n");
            writer.write("large_airport: " + largeAirport + "\n");
            writer.write("seaplane_base: " + seaplaneBase + "\n");
        } catch (Exception e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }

    }

    public static  void main(String[] args) {
        // Configurar Spark
        SparkConf sparkConf = new SparkConf().setAppName("Airports").setMaster("local[*]");
        JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);
        sparkContext.setLogLevel("ERROR");

        // Definir la ruta del archivo
        String pathFile = "./src/main/resources/airports.csv";

        // Leer el archivo como un RDD
        JavaRDD<String> rdd = sparkContext.textFile(pathFile);

        // Eliminar cabecera
        String header = rdd.first();
        JavaRDD<String[]> data = rdd
                .filter(line -> !line.equals(header))
                .map(line -> line.split(","));

        // Obtener índice de la columna "iso_country"

        // Limpiar los nombres de las columnas eliminando comillas
        List<String> columnNames = Arrays.stream(header.split(","))
                .map(s -> s.replaceAll("^\"|\"$", "")) // Elimina comillas dobles al inicio y al final
                .collect(Collectors.toList());

        // Obtener índice de la columna "iso_country"
        int indexIsoCountry = columnNames.indexOf("iso_country");
        int indexType = columnNames.indexOf("type");

        if (indexIsoCountry == -1 || indexType == -1) {
            System.err.println("Error: No se encontraron las columnas 'iso_country' o 'type' en el archivo CSV.");
            sparkContext.close();
            return;
        }

        // Filtrar aeropuertos de España
        JavaRDD<String[]> spanishAirports = data.filter(cols -> cols[indexIsoCountry].equals("ES"));

        // Contar cada tipo de aeropuerto
        long smallAirport = spanishAirports.filter(cols -> cols[indexType].contains("small_airport")).count();
        long heliport = spanishAirports.filter(cols -> cols[indexType].contains("heliport")).count();
        long mediumAirport = spanishAirports.filter(cols -> cols[indexType].contains("medium_airport")).count();
        long closed = spanishAirports.filter(cols -> cols[indexType].contains("closed")).count();
        long largeAirport = spanishAirports.filter(cols -> cols[indexType].contains("large_airport")).count();
        long seaplaneBase = spanishAirports.filter(cols -> cols[indexType].contains("seaplane_base")).count();

        System.out.println( "small_airport: " + smallAirport);
        System.out.println( "heliport: " + heliport);
        System.out.println( "medium_airport: " + mediumAirport);
        System.out.println( "closed: " + closed);
        System.out.println( "large_airport: " + largeAirport);
        System.out.println( "seaplane_base: " + seaplaneBase);

        writeResult(smallAirport, heliport, mediumAirport, closed, largeAirport, seaplaneBase);


        // Cerrar el contexto de Spark
        sparkContext.close();
    }
}
